pkg/vm/engine/tae/logtail/service/server.go

```diff
@@ -0,0 +1,465 @@
+		}
+	}
+}
+
+// logtailSender sends total or additional logtail.
+func (s *LogtailServer) logtailSender() {
+	logger := s.options.logger
+
+	for {
+		select {
+		case <-s.rootCtx.Done():
+			logger.Error("stop subscription handler", zap.Error(s.rootCtx.Err()))
+			return
+
+		case sub, ok := <-s.subChan:
+			if !ok {
+				logger.Info("subscription channel closed")
+				return
+			}
+
+			logger.Debug("start to handle subscription", zap.Any("table", sub.req.Table))
+
+			// if table subscribed already, we would ignore sub.req.CnWant.
+			want := s.subscribed.Waterline(sub.tableID, *sub.req.CnWant)
+
+			// fetch logtail
+			tail, err := s.fetcher.FetchLogtail(sub.sendCtx, sub.req.Table, nil, &want)
+			if err != nil {
+				logger.Error("fail to fetch logtail", zap.Error(err))
+
+				if err := sub.session.SendErrorResponse(
+					sub.sendCtx, *sub.req.Table, moerr.ErrInternal, "fail to fetch logtail",
+				); err != nil {
+					logger.Error("fail to send error response", zap.Error(err))
+				}
+
+				continue
+			}
+
+			// send subscription response
+			if err := sub.session.SendSubscriptionResponse(sub.sendCtx, tail); err != nil {
+				logger.Error("fail to send subscription response", zap.Error(err))
+				continue
+			}
+
+			// mark table as subscribed
+			sub.session.AdvanceState(sub.tableID)
+
+			// register subscribed table
+			s.subscribed.Register(sub.tableID, *sub.req.Table, want)
+
+		case pub, ok := <-s.pubChan:
+			if !ok {
+				logger.Info("publishment channel closed")
+				return
+			}
+
+			// publish all subscribed tables via session manager
+			for _, session := range s.ssmgr.ListSession() {
+				qualified := session.FilterLogtail(pub.tails...)
+				if err := session.SendUpdateResponse(s.rootCtx, qualified...); err != nil {
+					logger.Error("fail to publish additional logtail", zap.Error(err))
+					continue
+				}
+			}
+
+			// update waterline for all subscribed tables
+			s.subscribed.AdvanceWaterline(pub.want)
+		}
+	}
+}
+
+// collector collects logtail by interval.
+func (s *LogtailServer) collector() {
+	logger := s.options.logger
+
+	ticker := time.NewTicker(s.options.collectInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-s.rootCtx.Done():
+			logger.Error("stop logtail collector", zap.Error(s.rootCtx.Err()))
+			return
+
+		case <-ticker.C:
+			// take a snapshot for all subscribed tables
+			tables := s.subscribed.ListTable()
+
+			// take current timestamp
+			want := s.timestamp.Now()
+
+			tails := make([]tableLogtail, 0, len(tables))
+			for _, t := range tables {
+				tail, err := s.fetcher.FetchLogtail(s.rootCtx, &t.table, &t.waterline, &want)
+				if err != nil {
+					logger.Error("fail to fetch logtail", zap.Error(err), zap.Any("table", t.table))
+					continue
+				}
+				tails = append(tails, tableLogtail{id: t.id, tail: tail})
+			}
+			s.pubChan <- publishment{want: want, tails: tails}
+		}
+	}
+}
+
+// Close closes api server.
+func (s *LogtailServer) Close() error {
+	s.cancelFunc()
```

Comment:
you need wait all goroutine completed to avoid goroutine leak. You can use stopper to manager your async goroutines

Reply:
stopper would be better, I've update this PR in the following commit.